### PR TITLE
use libloading on osx to load renamex_np

### PR DESCRIPTION
**Pull request by debris**

### CODE REVIEW

File: src/platform/macos.rs [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
 use std::os::unix::ffi::OsStrExt;
 use std::path::Path;
 use std::{io, ffi};
+use self::libloading::os::unix::{Library, Symbol};
 
-extern "stdcall" {
-	fn renamex_np(oldpath: *const libc::c_char, newpath: *const libc::c_char, flags: libc::c_uint) -> libc::c_int;
+unsafe fn get_renamex_np() -> Option<Symbol<unsafe extern fn (oldpath: *const libc::c_char, newpath: *const libc::c_char, flags: libc::c_uint) -> libc::c_int>> {
+	let lib = Library::this();
+	lib.get(b"renamex_np").ok()
```
Comment (dvdplm) [inline src/platform/macos.rs]: Are we sure we don't want to panic here, `.expect("failed to get renamex_np, cannot proceed.")`?

File: src/platform/macos.rs [hunk 2]
```diff
@@ -17,13 +22,15 @@ pub fn swap<A, B>(a: A, b: B) -> io::Result<()> where A: AsRef<Path>, B: AsRef<P
...
 
 	unsafe {
-		if renamex_np(a_path.as_ptr(), b_path.as_ptr(), RENAME_SWAP) == 0 {
-			return Ok(());
-		}
-
-		let err = *libc::__error();
-		if err != libc::ENOTSUP {
-			return Err(io::Error::new(io::ErrorKind::Other, format!("renamex_np failed with code: {}", err)));
+		if let Some(renamex_np) = get_renamex_np() {
```
Comment (dvdplm) [inline src/platform/macos.rs]: …or handle the `None` here?

**Review by dvdplm — CHANGES_REQUESTED** [review]
Some error handling maybe?

File: src/platform/macos.rs (hunk 1 above)
Comment (debris) [reply src/platform/macos.rs]: we can proceed, `renamex_np` is not available only on osx < `10.12`. if it's not available, we fallback to `exchangedata`

**Comment by debris** [comment]
I believe that error handling is fine here

**Comment by debris** [comment]
maybe I'll add comments to the code

File: src/platform/macos.rs [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
+extern crate libloading;
 
 use std::os::unix::ffi::OsStrExt;
 use std::path::Path;
 use std::{io, ffi};
+use self::libloading::os::unix::{Library, Symbol};
 
-extern "stdcall" {
-	fn renamex_np(oldpath: *const libc::c_char, newpath: *const libc::c_char, flags: libc::c_uint) -> libc::c_int;
+unsafe fn get_renamex_np() -> Option<Symbol<unsafe extern fn (oldpath: *const libc::c_char, newpath: *const libc::c_char, flags: libc::c_uint) -> libc::c_int>> {
```
Comment (andresilva) [inline src/platform/macos.rs]: I have no idea how expensive this is. Do you think it makes sense to cache it?

File: src/platform/macos.rs [hunk 4]
```diff
@@ -1,12 +1,17 @@
...
+	let lib = Library::this();
+	lib.get(b"renamex_np").ok()
+}
 
+extern "stdcall" {
```
Comment (pepyakin) [inline src/platform/macos.rs]: wait, why "stdcall"? macOS uses system-v as a default calling convention.

File: src/platform/macos.rs (hunk 1 above)
Comment (dvdplm) [reply src/platform/macos.rs]: nice

**Comment by debris** [comment]
thanks everyone for the review! I merged and published the new version (`0.2.3`) to crates.io

File: src/platform/macos.rs (hunk 4 above)
Comment (andresilva) [reply src/platform/macos.rs]: Indeed, shouldn't this be failing? "stdcall" is only used on windows right?

Comment (pepyakin) [reply src/platform/macos.rs]: > "stdcall" is only used on windows right?

Yeah, that was my impression too.

Comment (debris) [reply src/platform/macos.rs]: ah, I missed your comment. This is not the change introduced in this pr. I probably just copy-pasted `extern xxx` from windows files. Yet, it's still interesting why it works

**Pull request merged**